# Advent of Code Day 1

In [124]:
#
from aocd import submit
from aocd.models import Puzzle

puzzle = Puzzle(day=1, year=2023)

# Get the imput and parse it
input = puzzle.input_data.split('\n')

# Using a lambda function to process each example
[(exampleA, answerA), (exampleB, answerB)] = [(ex.input_data.split('\n'), int(ex.answer_a or ex.answer_b)) for ex in puzzle.examples]

[(exampleA, answerA), (exampleB, answerB)]

[(['1abc2', 'pqr3stu8vwx', 'a1b2c3d4e5f', 'treb7uchet'], 142),
 (['two1nine',
   'eightwothree',
   'abcone2threexyz',
   'xtwone3four',
   '4nineeightseven2',
   'zoneight234',
   '7pqrstsixteen'],
  281)]

# First Part
The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover.

On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:

```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
```

In this example, the calibration values of these four lines are 12, 38, 15, and 77. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

In [125]:
def partA(input):
    result = 0
    for line in input:
        first = 0
        last = 0
        if line == "":
            continue
        for c in line:
            if not c.isdigit():
                continue
            if first == 0:
                first = c
            else:
                last = c

            if last == 0: 
                last = first
        x = int(first + last)
        result += x
    return result

[
    answerA == partA(exampleA),
    partA(input)
]

[True, 55447]

In [126]:
submit(partA(input), part="a", day=1, year=2023)

Part a already solved with same answer: 55447


# Second Part

Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:

```
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
```

In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. Adding these together produces 281.

In [194]:
import re

pattern = r"(one|two|three|four|five|six|seven|eight|nine|\d)"
numbers = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9"
}

def partB(input):
    result = 0
    for line in input:
        tokenized_line = re.findall(pattern, line)
        convert_tokens = [numbers.get(token, token) for token in tokenized_line]
        firstAndLast = convert_tokens[0] + convert_tokens[-1]
        if len(convert_tokens) >= 2:
            result += int(firstAndLast)
        else:
            result += int(convert_tokens[0])
            
        print(f"Line: {line}, Tokens: {convert_tokens}, FirstAndLast: {firstAndLast}, Sum: {result}")  # Debug print
        
    return result

# answerB == partB(exampleB)s
partB(["eightwo"])

Line: eightwo, Tokens: ['8'], FirstAndLast: 88, Sum: 8


8

In [191]:
partB(input)

51948

In [185]:
submit(partB(input), part="b", day=1, year=2023)

Line: kjrqmzv9mmtxhgvsevenhvq7, Tokens: ['9', '7', '7'], FirstAndLast: 97, Sum: 97
Line: four2tszbgmxpbvninebxns6nineqbqzgjpmpqr, Tokens: ['4', '2', '9', '6', '9'], FirstAndLast: 49, Sum: 146
Line: rkzlnmzgnk91zckqprrptnthreefourtwo, Tokens: ['9', '1', '3', '4', '2'], FirstAndLast: 92, Sum: 238
Line: fouronevzkbnzm6seven47, Tokens: ['4', '1', '6', '7', '4', '7'], FirstAndLast: 47, Sum: 285
[ERROR] Line with insufficient tokens: zphgdcznqsm2
Line: zphgdcznqsm2, Tokens: ['2'], FirstAndLast: 22, Sum: 287
Line: 4gjnmxtrbflgp71, Tokens: ['4', '7', '1'], FirstAndLast: 41, Sum: 328
Line: 4sqvv1cnpn, Tokens: ['4', '1'], FirstAndLast: 41, Sum: 369
Line: 8sevengzfvjrhnsb6ddb8ninerkgkxthtfkvbcmqs, Tokens: ['8', '7', '6', '8', '9'], FirstAndLast: 89, Sum: 458
Line: 1seven336, Tokens: ['1', '7', '3', '3', '6'], FirstAndLast: 16, Sum: 474
Line: 86one34vvvgdngbt39, Tokens: ['8', '6', '1', '3', '4', '3', '9'], FirstAndLast: 89, Sum: 563
Line: 37qdmsqzsq72clfntfxqfrhbxtmfourzcjxfmmfz, Tokens: ['3', '7'

wrong answer: That's not the right answer.  If you're stuck, make sure you're using the full input data; there are also some general tips on the about page, or you can ask for hints on the subreddit.  Because you have guessed incorrectly 4 times on this puzzle, please wait 5 minutes before trying again. [Return to Day 1]


That's not the right answer.  If you're stuck, make sure you're using the full input data; there are also some general tips on the about page, or you can ask for hints on the subreddit.  Because you have guessed incorrectly 4 times on this puzzle, please wait 5 minutes before trying again. [Return to Day 1]
